### ISSUE TITLE
CI fails with EADDRNOTAVAIL

### ISSUE THREAD

**Issue by cspramit** [casper-network/casper-node#3252]
Multiple tests are failing sporadically when an address cannot be acquired by the node or a client
example:
Casper-Test Nightly Test Status: Failure
Repo: casper-test
Author: marc-casperlabs
Drone Build: #512
Commit Link: ec2f21d5ff

**Comment by marc-casperlabs** [comment]
Removing this from 1.5.0, since it does not affect the release.

**Comment by marc-casperlabs** [comment]
Preliminary anaylsis:

We are getting some intermittent errors in CI and every now and then I get a bug report from that. These errors are usually hard to reproduce and I get a sneaking suspicion that they are related to overprovisioning on Drone's end.

As an example, we recently had a failure that I am looking at now (https://github.com/casper-network/casper-node/issues/3252), which looks like this:

> error sending request for url (http://localhost:11101/rpc): error trying to connect: tcp connect error: Cannot assign requested address (os error 99)

This (`EADDRNOTAVAIL` = 99) indicates that a client could not create an "outgoing" (they're all loopback) TCP connection because no ephemeral source port could be allocated for the initiating socket. Usually a few thousand of these are available, but they can get scarce if:

* Multiple things are running on a shared interface, taking up ports,
* the range of available ports has been reduced (e.g. we'd need to check what's in `/proc/sys/net/ipv4/ip_local_port_range`),
* older jobs have been run on the interface that still have lingering (as in `TIME_WAIT` or similar) connections,
* we're plain running too many things at once (maybe a high refresh rate requiring TCP connections, but we'd have seen that locally as well), or
* something's just plain broken on the host.

@dwerner pointed out that there is no reusage of `reqwest::Client` instances, which may be the underlying issue. Fixing that may already solve this problem.

**Comment by georgepisaltu** [comment]
The `EADDRNOTAVAIL` error must come from a `connect` syscall as the other syscalls which can yield this error are N/A (`bind` for example would error with `EADDRNOTAVAIL` if a nonexistent interface was provided, which is not the case for us). After looking at the node code, the `EADDRNOTAVAIL` error couldn't have come from there as the test would have exited with some other error (like `ECONNREFUSED` if the node crashed and nobody was listening on that address).

This means the bug is coming from somewhere inside the testing stack (NCTL, casper-test, casper-client). From the error message, we know it was yielded by casper-test when attempting an RPC call.

The man explanation for `EADDRNOTAVAIL` on [connect](https://man7.org/linux/man-pages/man2/connect.2.html) is:
 >            (Internet domain sockets) The socket referred to by sockfd
 >            had not previously been bound to an address and, upon
 >            attempting to bind it to an ephemeral port, it was
 >            determined that all port numbers in the ephemeral port
 >            range are currently in use.  See the discussion of
 >            /proc/sys/net/ipv4/ip_local_port_range in [ip(7)](https://man7.org/linux/man-pages/man7/ip.7.html).

The bug never reproduced locally, so it must have something to do with limited resources on the drone testing machine. This coupled with the fact that the `Client` instances were not reused meant it was very feasible for the drone machine to run out of resources, as pointed out by @dwerner and @marc-casperlabs respectively. My hunch is that we reached a point where we do so many requests in a normal test-all casper-test session that we reach the port exhaustion limit.

PRs for [casper-test](https://github.com/casper-network/casper-test/pull/104) and [casper-client](https://github.com/casper-ecosystem/casper-client-rs/pull/39) are up fixing the `reqwest::Client` usage in both repositories. This hopefully fixes the problem and confirms the hypothesis above.

To summarize, a node problem is ruled out, most likely the testing machine is running out of ports because of inefficient `reqwest::Client` usage.

**Comment by georgepisaltu** [comment]
The diagnostic script running after failures in `casper-test` showed that the port range was vast. Moreover, when testing locally only around 200 ports were used. These findings exclude the possibility of port exhaustion.

> The bug never reproduced locally, so it must have something to do with limited resources on the drone testing machine.
@TomVasile suggested we run the test locally, but inside a docker container, mimicking the drone setup. This made the bug reproduce almost all of the time. A sample `netstat` output shows that some nodes do not open their RPC ports by the time the 15 second timeout expires. This is because said nodes take longer to reach the participating state. Excerpt `netstat -tunap` output after a failed test:
```
tcp        0      0 0.0.0.0:14105           0.0.0.0:*               LISTEN      31407/casper-node   
tcp        0      0 0.0.0.0:14104           0.0.0.0:*               LISTEN      31405/casper-node   
tcp        0      0 0.0.0.0:14103           0.0.0.0:*               LISTEN      31380/casper-node   
tcp        0      0 0.0.0.0:14102           0.0.0.0:*               LISTEN      31353/casper-node   
tcp        0      0 0.0.0.0:14101           0.0.0.0:*               LISTEN      31354/casper-node   
tcp        0      0 0.0.0.0:22105           0.0.0.0:*               LISTEN      31407/casper-node   
tcp        0      0 0.0.0.0:22104           0.0.0.0:*               LISTEN      31405/casper-node   
tcp        0      0 0.0.0.0:22103           0.0.0.0:*               LISTEN      31380/casper-node   
tcp        0      0 0.0.0.0:22102           0.0.0.0:*               LISTEN      31353/casper-node   
tcp        0      0 0.0.0.0:22101           0.0.0.0:*               LISTEN      31354/casper-node   
tcp        0      0 0.0.0.0:11103           0.0.0.0:*               LISTEN      31380/casper-node   
tcp        0      0 0.0.0.0:11102           0.0.0.0:*               LISTEN      31353/casper-node   
tcp        0      0 0.0.0.0:11105           0.0.0.0:*               LISTEN      31407/casper-node   
tcp        0      0 0.0.0.0:11104           0.0.0.0:*               LISTEN      31405/casper-node   
tcp        0      0 0.0.0.0:18105           0.0.0.0:*               LISTEN      31407/casper-node   
tcp        0      0 0.0.0.0:18104           0.0.0.0:*               LISTEN      31405/casper-node   
tcp        0      0 0.0.0.0:18103           0.0.0.0:*               LISTEN      31380/casper-node   
tcp        0      0 0.0.0.0:18102           0.0.0.0:*               LISTEN      31353/casper-node   
tcp        0      0 0.0.0.0:18101           0.0.0.0:*               LISTEN      31354/casper-node
```

Now that we know the client connect error is justified since there sometimes isn't anything to connect to, we have two distinct problems that only affect us inside the docker container:
1. Why are some nodes now taking longer to reach the participating state?
2. Why is the error we get `EADDRNOTAVAIL` instead of `ECONNREFUSED`?

For the first problem, the time difference on average for reaching participating state on the laziest node _on my machine_ is 25s inside the container and 9s on the host. Another problem why did it start happening only about a month ago. It would be very difficult to study container performance compared to native and it is most likely related to networking. I tried following the process with `strace` and the process sleeps a lot for 100ms, meaning it probably does `fetch_and_store_next` a lot and runs out of peers, which would indicate the network is slow since the NTCL/test requirements for reaching the participating state haven't changed recently.

The second problem is harder to diagnose and would require a deep dive into the linux network stack code and docker's setup as well. Since at the time the test failed there was no RPC port listened to (as `netstat` would show), all documentation points to an `ECONNREFUSED` error. The `man` description is:
> A `connect()` on a stream socket found no one listening on the remote address.

Other people have signaled issues regarding unexpected `EADDRNOTAVAIL` errors on `connect` when containerizing their applications using docker. Also, we only get this error inside the container.

At this point, I'd argue the first problem is more important than the second and we should focus on that. In local test runs, an increase to the node readiness timeout from 15s to 30s gives the nodes enough time to reach the participating state and fixes the issue. Fully understanding what caused this is yet to be determined.

**Comment by georgepisaltu** [comment]
The initial issue seems to be fixed, but now the [CI fails at another point](https://drone-auto-casper-network.casperlabs.io/casper-network/casper-test/558/1/2), namely when running the `single node join` test spec. Messages like the following are logged every 5 seconds and then the test fails after 5 minutes:
> 2022-09-04T05:44:20.111731Z  WARN process{trigger=ensure did not crash, on all nodes}: casper_test::spec::checks: node reports stopped (but not crashed) node=nctl-node 0

This might be because the time is still not enough for one node to reach participating and it could never recover from that state, therefore compromising the testing scenario
> 2022-09-04T05:44:20.102840Z  WARN casper_test::environment::nctl: not all nctl nodes reported readiness after timeout failed_nodes=nctl-node 4 timeout=30s

We can try to increase the timeout once again as the previous figure of 15s -> 30s was based on observations on my local machine and it might not be suitable for the drone environment.

**Comment by georgepisaltu** [comment]
The original issue was fixed by https://github.com/casper-network/casper-test/pull/108. The second problem of the failing `simple_join` test was fixed by @TomVasile in https://github.com/casper-network/casper-test/pull/109. The test are running smoothly. Closing the issue.
